### PR TITLE
Detect sysroot automatically

### PR DESCRIPTION
**Pull request by chrisburr**
### Setting the sysroot

This PR tries to fix the sysroot situation so it always does the correct thing. See https://github.com/conda-forge/clangdev-feedstock/pull/45#issuecomment-457223530 for the logic that is used. It probably makes sense to cherry pick these patches in to the other branches when we're happy with them.

I've tested building ROOT and cling and this will fix https://github.com/conda-forge/root-feedstock/issues/2, https://github.com/conda-forge/cling-feedstock/issues/18#issuecomment-455911070 and https://github.com/QuantStack/xeus-cling/issues/194#issuecomment-456547091.

### New patches (WIP)

A new release of ROOT has now been made so we need some more patches. I've also asked for a new cling release so these patches will be needed for that too. (With the aim of modifying the ROOT feedstock to use an external build of cling.)

Rather than constantly needing to keep the `clangdev` patch list up to date, I'm tempted to pull the sources from [ROOT's fork](https://github.com/root-project/root/tree/master/interpreter/llvm/src/tools/clang).

### CODE REVIEW

**Comment by chrisburr** [comment]
@SylvainCorlay Can you take a look at why the macOS build is failing? I think it's been caused by the compiler switch as it also affects the main `5.x` branch.

For the `default` variant:
```
RuntimeError: clangdev cannot depend on itself
```
For the `cling` variant:
```
conda_build.exceptions.RecipeError: Mismatching hashes in recipe. Exact pins in dependencies that contribute to the hash often cause this. Can you change one or more exact pins to version bound constraints?
Involved packages were:
Mismatching package: clang_variant (id cling); dep: clang_variant 1.0 default; consumer package: clangdev
```

**Comment by SylvainCorlay** [comment]
The relabelling of gcc7-labelled packages seems to have created duplicates, which probably have different hashes:

screenshot from 2019-01-17 18-48-05

**Comment by SylvainCorlay** [comment]
This probably occured with other mutex packages that have build strings without hashes, like the MPI one (cc @minrk).

**Comment by chrisburr** [comment]
I'm finally getting somewhere with why this fails. For some reason `llvmdev=5.0.0=h04f5b5a_1001` depends on `clangdev`. In my previous PR `llvmdev=5.0.0=h2d50403_1` was used which doesn't have a dependency on `clangdev`.

(This only fixes the `clangdev cannot depend on itself` error.)

File: recipe/build_clangdev.sh [hunk 1]
```diff
@@ -1,17 +1,16 @@
 if [[ "$(uname)" == "Linux" && "$cxx_compiler" == "gxx" ]]; then
-    sed -i.bak -e 's@addSystemInclude(DriverArgs, CC1Args, SysRoot + "/usr/local/include");@addSystemInclude(DriverArgs, CC1Args, "'"${PREFIX}/${HOST}/sysroot/usr/include"'");@g' \
```
Comment (chrisburr) [inline recipe/build_clangdev.sh]: Note to myself for later: `SysRoot + "/usr/local/include"` needs to be replaced with `SysRoot + "/usr/include"`.

**Comment by chrisburr** [comment]
@conda-forge-admin, please rerender

**Comment by chrisburr** [comment]
Logic for finding the sysroot (assuming `-isysroot` isn't passed):

Linux:
 1. Use the `CONDA_BUILD_SYSROOT` env variable if it is set and is an absolute path that exists
 2. Fall back to using `$PREFIX/$HOST/sysroot`

macOS:
 1. Use the `CONDA_BUILD_SYSROOT` env variable if it is set, is an absolute path that exists and isn't `/`.
 2. Use the `SDKROOT` env variable if it is set, is an absolute path that exists and isn't `/`.
 3. Use `/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk` if it contains `/usr/include/sys/types.h`
 4. Use `/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk` if it contains `/usr/include/sys/types.h`
 5. Use `/` if `/usr/include/sys/types.h` exists (i.e. macOS <= 10.13)
 6. Run `xcode-select -p` and check if the command line tools are installed in a non standard location
 6. Run `xcode-select -p` and check if xcode is installed in a non standard location

**Comment by chrisburr** [comment]
@SylvainCorlay I've edited the PR description to explain the current state of this, can you take a look?

**Comment by SylvainCorlay** [comment]
> ### Setting the sysroot
> This PR tries to fix the sysroot situation so it always does the correct thing. See [#45 (comment)](https://github.com/conda-forge/clangdev-feedstock/pull/45#issuecomment-457223530) for the logic that is used.
> 
> This will fix [conda-forge/root-feedstock#2](https://github.com/conda-forge/root-feedstock/issues/2), [conda-forge/cling-feedstock#18 (comment)](https://github.com/conda-forge/cling-feedstock/issues/18#issuecomment-455911070) and [QuantStack/xeus-cling#194 (comment)](https://github.com/QuantStack/xeus-cling/issues/194#issuecomment-456547091).
> 

Excellent!

> ### New patches (WIP)
> A new release of ROOT has now been made so we need some more patches. I've also asked for a new cling release so these patches will be needed for that too. (With the aim of modifying the ROOT feedstock to use an external build of cling.)
> 

Great!

 - Should we add the new patches in a separate PR?
 - do these patches modify the clang public API & ABI?

> Rather than constantly needing to keep the `clangdev` patch list up to date, I'm tempted to pull the sources from [ROOT's fork](https://github.com/root-project/root/tree/master/interpreter/llvm/src/tools/clang).

The issue with ROOT's fork is that they apply patches on the top of an arbitrary commit of the clang history instead of starting off a release tag, making it so that we cannot apply the patches to a release source tarball.

In the case of this repository, I actually rebased the patches on the top of a release tarball!

I have been trying to communicate that issue with the ROOT team with little success, although it makes the whole thing very hard to reproduce.

**Comment by SylvainCorlay** [comment]
> The issue with ROOT's fork is that they apply patches on the top of an arbitrary commit of the clang history instead of starting off a release tag, making it so that we cannot apply the patches to a release source tarball.
>
> In the case of this repository, I actually rebased the patches on the top of a release tarball!
>
> I have been trying to communicate that issue with the ROOT team with little success, although it makes the whole thing very hard to reproduce.
>

**Pull request closed without merging**